### ISSUE TITLE
feat(lab): import match data from ipscresults.org OData API

### ISSUE THREAD

**Issue by mandakan** [mandakan/ssi-scoreboard#225]
## Summary

Expand the ratings lab dataset by importing match results from [ipscresults.org](https://ipscresults.org), the official IPSC results repository. This adds ~1,254 L3–L5 matches (2009–present, worldwide) to complement the existing SSI dataset (~222 matches, mostly Scandinavian L2–L3, 2021–present).

ipscresults.org exposes a **public OData JSON API** — no authentication required, no scraping needed.

### API endpoints (verified working)

| Endpoint | Returns |
|---|---|
| `StatsMatchList` | All matches: ID (UUID), Name, Date, Level (3–5), Region, Discipline |
| `StatsMatchDetail({id})` | Match detail: region code, location, match director |
| `Stats.DivisionList(id)` | Divisions with competitor counts per category |
| `Stats.CompetitorList(id)` | All competitors: name, alias, region, division, category, DQ |
| `Stats.StageList(id, div)` | Stages: name, course type, target counts, max points |
| `Stats.StageResult(id, div)` | Per-stage results: rank, name, time, score, hit factor, stage percent |
| `Stats.CompetitorScore(id, div, comp)` | Per-stage A/C/D hits, misses, penalties (not needed initially, see #future) |

Full OData metadata: `https://ipscresults.org/odata/$metadata`

---

## Design principles

- **Sync sources independently, link/dedup as a separate step.** Each source syncs into its own rows tagged by `source`. Identity resolution and match deduplication run locally on already-synced data — no network calls needed.
- **Source-agnostic algorithms.** The training step sees only `canonical_shooter_id` — it doesn't know or care where the data came from.
- **Extensible for future sources.** PractiScore or others follow the same pattern: add a sync client, plug into the same identity/dedup infrastructure.

---

## Phase 1: Schema changes (`store.py`)

Add `source TEXT` column to all four data tables. Change PKs to include `source`:

```
matches:       PK (source, ct, match_id)
competitors:   PK (source, ct, match_id, competitor_id)
stages:        PK (source, ct, match_id, stage_id)
stage_results: PK (source, ct, match_id, competitor_id, stage_id)
```

Add two new tables for identity resolution and match deduplication:

```sql
-- Canonical shooter identity across sources
shooter_identities (
  canonical_id   INTEGER PRIMARY KEY AUTOINCREMENT,
  primary_name   TEXT,
  region         TEXT,
  created_at     TIMESTAMP
);

-- Links a source-specific identity to a canonical shooter
shooter_identity_links (
  source         TEXT,        -- 'ssi' | 'ipscresults'
  source_key     TEXT,        -- ssi: shooter_id; ipscresults: normalized "last,first|REG"
  canonical_id   INTEGER REFERENCES shooter_identities,
  name_variant   TEXT,        -- exact name as seen in this source
  confidence     REAL,        -- 1.0 = exact, 0.0–1.0 = fuzzy
  method         TEXT,        -- 'auto_exact', 'auto_fuzzy', 'manual'
  linked_at      TIMESTAMP,
  PRIMARY KEY (source, source_key)
);

-- Match deduplication links
match_links (
  source_a       TEXT,
  ct_a           INTEGER,
  match_id_a     TEXT,
  source_b       TEXT,
  ct_b           INTEGER,
  match_id_b     TEXT,
  confidence     REAL,
  method         TEXT,        -- 'auto_name_date', 'auto_roster', 'manual'
  preferred      TEXT,        -- 'a' or 'b' — which copy has better data
  linked_at      TIMESTAMP,
  PRIMARY KEY (source_a, match_id_a, source_b, match_id_b)
);
```

Default `source='ssi'` for all existing data. Since we can easily re-read all data, table recreation on schema version bump is acceptable.

---

## Phase 2: ipscresults OData client (`src/data/ipscresults.py`)

Two classes:

### `IpscResultsClient` — thin OData HTTP wrapper
- `get_match_list(level_min, discipline)` — paginated via `$skip/$top`
- `get_match_detail(match_id)`
- `get_divisions(match_id)`
- `get_stage_list(match_id, div_code)`
- `get_stage_results(match_id, div_code)`
- Rate limiting: configurable delay + jitter (same pattern as `SyncClient`)
- All return Pydantic models (new `src/data/ipscresults_models.py`)

### `IpscResultsSyncer` — converts OData → `MatchResults` → `Store`
- For each match: fetch divisions → for each division fetch stage list + stage results
- Map to existing `MatchResults` / `StageResult` models
- Name normalization: `"Last, First"` → `"First Last"`
- Level mapping: `3` → `"l3"`, `4` → `"l4"`, `5` → `"l5"`
- `competitor_id`: use ipscresults `CompetitorNumber` (unique within match)
- `shooter_id`: leave as `None` — resolved later by identity linking
- `source`: always `"ipscresults"`, `ct`: `0`

### Field mapping

| ipscresults field | Lab field | Notes |
|---|---|---|
| `StageResult.HitFactor` | `hit_factor` | Direct |
| `StageResult.Score` | `points` | Direct |
| `StageResult.StageTime` | `time` | Direct |
| `StageList.MaxPoints` | `max_points` | From stage metadata |
| `StageResult.Rank` | `division_rank` | Per-division only |
| `StageResult.StagePercent` | `division_percent` | Per-division only |
| `CompetitorList.DQ` | `dq` | Boolean, applied to all stages |
| A/C/D hits, misses, procedurals | `NULL` | Requires `CompetitorScore` endpoint (future improvement) |
| `overall_rank`, `overall_percent` | Computed | Can be computed locally from cross-division HF data |

---

### Name normalization (`normalize_name`)
1. Source-aware parsing: ipscresults `"Last, First"` → `"first last"`, SSI `"First Last"` → `"first last"`
2. Strip diacritics for comparison: `"Sjöberg"` → `"sjoberg"` (via `unicodedata.normalize`)
3. Handle edge cases from SSI data: `"Kenneth NotKnown"`, single-initial names like `"Adrian H"`
4. Generate fingerprint: `lowercase(first_last) + "|" + region_code`

### Linking workflow (`resolve_identities`)
1. **Bootstrap SSI identities:** Each unique `(shooter_id, name, region)` → `shooter_identities` row + link. Handle name variants (e.g. `['Marianne Schön', 'Marianne Hansen']` for same shooter_id) by using most recent name as `primary_name` and registering all variants as fingerprints.
2. **Match ipscresults competitors:**
   - Exact fingerprint match → `confidence=1.0`, `method='auto_exact'`
   - Levenshtein distance ≤ 2, same region → `confidence=0.8`, `method='auto_fuzzy'`
   - No match → new `canonical_id`
3. **Manual overrides:** `rating link-shooter` CLI command, stored with `method='manual'`, never overwritten by auto-resolution.

### Known identity challenges (from SSI data)

Shooters with name changes, abbreviations, nicknames, or diacritics variations:
```
shooter_id=14150  ['Jack af Janérs', 'Jack Janérs', 'One-Eyed Jack']
shooter_id=45199  ['jim sæther', 'Jim-Rune Saether', 'jim-rune sæther']
shooter_id=715    ['Sasa Petrovic', 'Saša Petrović']
shooter_id=1679   ['Marianne Schön', 'Marianne Hansen']  (last name change)
```

---

### `find_duplicate_matches` — heuristics in order of confidence:
1. **Name + date exact:** Fuzzy name match + same date → `confidence=0.95`
2. **Name similarity + date proximity:** Levenshtein ratio > 0.8, date within ±3 days → `confidence=0.7`
3. **Roster overlap:** >60% competitor fingerprints match between matches on same date → `confidence=0.9`

### `apply_dedup_for_training` → `set[(source, ct, match_id)]`
Returns matches to skip. For each linked pair, keep the `preferred` side (whichever has more stages/competitors/complete data).

Overlap window: L3 Scandinavian matches from 2021 onward exist on both SSI and ipscresults.

---

## Phase 5: Training pipeline integration

1. `get_matches_chronological()` returns `(source, ct, match_id, date, level)` tuples
2. `get_competitor_shooter_map()` extended to return `canonical_id` via identity links
3. Before training loop, filter out dedup skip set
4. Rating algorithms need **zero changes** — they see `int` shooter IDs as before

---

## Phase 6: CLI commands

```
rating sync              # existing SSI sync (unchanged, adds source='ssi')
rating sync-ipscresults  # new: sync from ipscresults.org OData API
  --level-min 3          # minimum match level (default 3)
  --discipline Handgun   # filter (default Handgun)
  --full                 # ignore watermark
  --delay 0.5            # rate limit

rating link              # run identity resolution + match dedup
  --force                # re-run even if already linked
  --report               # print summary of linked/fuzzy/unlinked counts

rating link-shooter      # manual identity merge
  --canonical-id 42
  --source ipscresults
  --source-key "hollertz, martin|SWE"

rating train             # unchanged, now uses canonical IDs and skips dupes
rating pipeline          # extended: sync → sync-ipscresults → link → train → export
```

---

## Phase 7: Tests

| Test file | Covers |
|---|---|
| `tests/test_identity.py` | Name normalization, fingerprinting, exact/fuzzy matching, name variants, manual overrides |
| `tests/test_match_dedup.py` | Name+date matching, roster overlap, preferred source selection |
| `tests/test_ipscresults.py` | OData response parsing, field mapping, level mapping |
| `tests/test_store.py` | Extended: source column, identity/dedup tables, canonical shooter maps |

---

## File layout (new/modified)

```
lab/src/data/
├── models.py                # Modified: add source field to MatchResultsMeta
├── store.py                 # Modified: source column, identity + dedup tables
├── sync.py                  # Modified: pass source='ssi'
├── ipscresults.py           # NEW: OData client + syncer
├── ipscresults_models.py    # NEW: Pydantic models for OData responses
├── identity.py              # NEW: name normalization + identity resolution
├── match_dedup.py           # NEW: match deduplication logic
└── exporter.py              # Modified: use canonical IDs
```

---

## Execution order

Each phase is independently testable and committable:
1. Schema changes + tests (Phase 1 + 7)
2. ipscresults client + tests (Phase 2 + 7)
3. Identity resolution + tests (Phase 3 + 7)
4. Match dedup + tests (Phase 4 + 7)
5. Training integration (Phase 5)
6. CLI wiring (Phase 6)

---

## Data landscape

| Source | Matches | Date range | Levels | Region focus |
|---|---|---|---|---|
| SSI (current) | ~222 | 2021–2026 | L2–L3 | Scandinavia |
| ipscresults.org | ~1,254 | 2009–present | L3–L5 | Worldwide |
| Combined | ~1,450+ | 2009–2026 | L2–L5 | Mixed |

The combined dataset gives ~6× more matches, covers higher-level competitions (continentals, world shoots), and spans a much longer time window — critical for testing inactivity decay and level-scaling algorithms.
